src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.TypeSystem;
+using System;
+using System.Net;
+
+namespace Internal.TypeSystem
+{
+    public class ExplicitLayoutValidator
+    {
+        private enum FieldLayoutTag : byte
+        {
+            Empty,
+            NonORef,
+            ORef,
+        }
+
+        private readonly int _pointerSize;
+
+        private readonly FieldLayoutTag[] _fieldLayout;
+
+        private readonly MetadataType _typeBeingValidated;
+
+        private ExplicitLayoutValidator(MetadataType type, int typeSizeInBytes)
+        {
+            _typeBeingValidated = type;
+            _pointerSize = type.Context.Target.PointerSize;
+            _fieldLayout = new FieldLayoutTag[typeSizeInBytes];
+        }
+
+        public static void Validate(MetadataType type, ComputedInstanceFieldLayout layout)
+        {
+            ExplicitLayoutValidator validator = new ExplicitLayoutValidator(type, layout.ByteCountUnaligned.AsInt);
+            foreach (FieldAndOffset fieldAndOffset in layout.Offsets)
+            {
+                validator.AddToFieldLayout(fieldAndOffset.Offset.AsInt, fieldAndOffset.Field.FieldType);
+            }
+        }
+
+        private void AddToFieldLayout(int offset, TypeDesc fieldType)
+        {
+            if (fieldType.IsGCPointer)
+            {
+                if (offset % _pointerSize != 0)
+                {
+                    // Misaligned ORef
+                    ThrowFieldLayoutError(offset);
+                }
+                SetFieldLayout(offset, _pointerSize, FieldLayoutTag.ORef);
+            }
+            else if (fieldType.IsValueType)
+            {
+                MetadataType mdType = (MetadataType)fieldType;
+                int fieldSize = mdType.InstanceByteCountUnaligned.AsInt;
+                if (!mdType.ContainsGCPointers)
+                {
+                    // Plain value type, mark the entire range as NonORef
+                    SetFieldLayout(offset, fieldSize, FieldLayoutTag.NonORef);
+                }
+                else
+                {
+                    if (offset % _pointerSize != 0)
+                    {
+                        // Misaligned struct with GC pointers
+                        ThrowFieldLayoutError(offset);
+                    }
+
+                    bool[] fieldORefMap = new bool[fieldSize];
+                    MarkORefLocations(mdType, fieldORefMap, offset: 0);
+                    for (int index = 0; index < fieldSize; index++)
+                    {
+                        SetFieldLayout(offset + index, fieldORefMap[index] ? FieldLayoutTag.ORef : FieldLayoutTag.NonORef);
+                    }
+                }
+            }
+            else if (fieldType.IsPointer || fieldType.IsByRef || fieldType.IsByRefLike)
```

Comment:
* IsByRefLike is a ref struct - those will respond to `IsValueType` above.
* We don't allow ByRef fields right now, so maybe we don't need to handle that.
* Pointer typed fields are okay to be misaligned. This works:

```csharp
[StructLayout(LayoutKind.Explicit)]
unsafe class Program
{
    [FieldOffset(2)] int* f;

    static unsafe void Main()
    {
    }
}
```

Reply:
Removed IsByRefLike from the conditional check. Added a new clause for IsPointer setting the range to NonORef.